tests/subsys/modem/backends/uart/src/main.c

```diff
@@ -0,0 +1,212 @@
+static uint32_t transmit_prng_state = 1234;
+static uint32_t receive_prng_state = 1234;
+static uint32_t transmit_size_prng_state;
+
+static uint8_t transmit_prng_random(void)
+{
+	transmit_prng_state = ((1103515245 * transmit_prng_state) + 12345) % (1U << 31);
+	return (uint8_t)(transmit_prng_state & 0xFF);
+}
+
+static uint8_t receive_prng_random(void)
+{
+	receive_prng_state = ((1103515245 * receive_prng_state) + 12345) % (1U << 31);
+	return (uint8_t)(receive_prng_state & 0xFF);
+}
+
+static void prng_reset(void)
+{
+	transmit_prng_state = 1234;
+	receive_prng_state = 1234;
+	transmit_size_prng_state = 0;
+}
+
+static void fill_transmit_ring_buf(void)
+{
+	uint32_t space = ring_buf_space_get(&transmit_ring_buf);
+	uint8_t data;
+
+	for (uint32_t i = 0; i < space; i++) {
+		data = transmit_prng_random();
+		ring_buf_put(&transmit_ring_buf, &data, 1);
+	}
+}
+
+static uint32_t transmit_size_prng_random(void)
+{
+	uint32_t size = 1;
+
+	for (uint8_t i = 0; i < transmit_size_prng_state; i++) {
+		size = size * 2;
+	}
+
+	transmit_size_prng_state = transmit_size_prng_state == 11
+				 ? 0
+				 : transmit_size_prng_state + 1;
+
+	return size;
+}
+
+static int transmit_prng(uint32_t remaining)
+{
+	uint8_t *reserved;
+	uint32_t reserved_size;
+	uint32_t transmit_size;
+	int ret;
+
+	fill_transmit_ring_buf();
+	reserved_size = ring_buf_get_claim(&transmit_ring_buf, &reserved, UINT32_MAX);
+	transmit_size = MIN(transmit_size_prng_random(), reserved_size);
+	transmit_size = MIN(remaining, transmit_size);
+	ret = modem_pipe_transmit(pipe, reserved, transmit_size);
+	if (ret < 0) {
+		return ret;
+	}
+	printk("TX: %u,%u\n", transmit_size, (uint32_t)ret);
+	__ASSERT(ret <= remaining, "Impossible number of bytes sent %u", (uint32_t)ret);
+	ring_buf_get_finish(&transmit_ring_buf, ret);
+	return ret;
+}
+
+static int receive_prng(void)
+{
+	int ret = 0;
+
+	if (k_sem_take(&receive_ready_sem, K_NO_WAIT) == 0) {
+		ret = modem_pipe_receive(pipe, receive_buffer, sizeof(receive_buffer));
+		if (ret < 0) {
+			return -EFAULT;
+		}
+		for (uint32_t i = 0; i < (uint32_t)ret; i++) {
+			if (receive_prng_random() != receive_buffer[i]) {
+				return -EFAULT;
+			}
+		}
+		printk("RX: %u\n", (uint32_t)ret);
+	}
+
+	return ret;
+}
+
+/*************************************************************************************************/
+/*                                         Test setup                                            */
+/*************************************************************************************************/
+static void *test_modem_backend_uart_setup(void)
+{
+	const struct modem_backend_uart_config config = {
+		.uart = uart,
+		.receive_buf = backend_receive_buffer,
+		.receive_buf_size = 1024,
+		.transmit_buf = backend_transmit_buffer,
+		.transmit_buf_size = 1024,
+	};
+
+	pipe = modem_backend_uart_init(&uart_backend, &config);
+	modem_pipe_attach(pipe, modem_pipe_callback_handler, NULL);
+	k_msleep(10);
```

Comment:
I never like time delays in code. Do we have to have this? If so, add a comment why.

Reply:
This was a leftover from me trying to solve this bug #65320, I have removed the timeout :)